test/common/quic/envoy_quic_h3_fuzz_test.cc

```diff
@@ -0,0 +1,304 @@
+                          quic::ConnectionCloseSource source) override {
+    quic::QuicDispatcher::OnConnectionClosed(connection_id, error, error_details, source);
+  }
+  quic::QuicTimeWaitListManager* CreateQuicTimeWaitListManager() override {
+    return new EnvoyQuicTimeWaitListManager(writer(), this, helper()->GetClock(), alarm_factory(),
+                                            quic_stats_);
+  }
+
+  void closeConnectionsWithFilterChain(const Network::FilterChain*) {}
+
+  void updateListenerConfig(Network::ListenerConfig& new_listener_config) {
+    listener_config_ = &new_listener_config;
+  }
+
+protected:
+  std::unique_ptr<quic::QuicSession>
+  CreateQuicSession(quic::QuicConnectionId server_connection_id,
+                    const quic::QuicSocketAddress& self_address,
+                    const quic::QuicSocketAddress& peer_address, absl::string_view /*alpn*/,
+                    const quic::ParsedQuicVersion& version,
+                    const quic::ParsedClientHello& /*parsed_chlo*/) override {
+
+    auto connection_socket = Quic::createConnectionSocket(peer_addr_, self_addr_, nullptr);
+    auto connection = std::make_unique<EnvoyQuicServerConnection>(
+        server_connection_id, self_address, peer_address, *helper(), *alarm_factory(), writer(),
+        false, quic::ParsedQuicVersionVector{quic_version_}, std::move(connection_socket),
+        connection_id_generator());
+
+    auto decrypter = std::make_unique<FuzzDecrypter>();
+    auto encrypter = std::make_unique<quic::NullEncrypter>(quic::Perspective::IS_CLIENT);
+    connection->InstallDecrypter(quic::EncryptionLevel::ENCRYPTION_FORWARD_SECURE,
+                                 std::move(decrypter));
+    connection->SetEncrypter(quic::EncryptionLevel::ENCRYPTION_FORWARD_SECURE,
+                             std::move(encrypter));
+
+    connection->SetDefaultEncryptionLevel(quic::EncryptionLevel::ENCRYPTION_FORWARD_SECURE);
+
+    auto stream_info = std::make_unique<StreamInfo::StreamInfoImpl>(
+        dispatcher_.timeSource(),
+        connection->connectionSocket()->connectionInfoProviderSharedPtr());
+    auto session = std::make_unique<EnvoyQuicServerSession>(
+        quic_config_, quic::ParsedQuicVersionVector{version}, std::move(connection), this,
+        &crypto_stream_helper_, &crypto_config_, &compressed_certs_cache_, dispatcher_,
+        quic::kDefaultFlowControlSendWindow * 1.5, quic_stat_names_,
+        listener_config_->listenerScope(), crypto_stream_factory_, std::move(stream_info),
+        connection_stats_);
+    session->Initialize();
+    session->setHeadersWithUnderscoreAction(envoy::config::core::v3::HttpProtocolOptions::ALLOW);
+    session->setHttp3Options(http3_options_);
+    session->setCodecStats(http3_stats_);
+    session->setHttpConnectionCallbacks(http_connection_callbacks_);
+    session->setMaxIncomingHeadersCount(100);
+    session->set_max_inbound_header_list_size(64 * 1024u);
```

Comment:
The fuzzer targets the codec libraries. Adding HCM would incur a performance hit for it will allocate streams. Further, HCM is protocol agnostic, is there a good reason to fuzz quic/h3 specifically and HCM at the same time?

Reply:
QUICHE has its own fuzzer already, so I would think the value of this one is to fuzz its interaction with HCM. But I could be wrong about what this PR is supposed to achieve. If Envoy is concerned about the safety of external libraries, it's totally fine to add its own fuzzer for QUICHE.

Reply:
There is more interaction, c.f. https://github.com/envoyproxy/envoy/pull/26573
This fuzzer is part of an ongoing effort to shield the envoy code base from erroneous commits.